aes/src/autodetect.rs

```diff
@@ -0,0 +1,204 @@
+            #[inline]
+            fn encrypt_block(&self, block: &mut Block) {
+                if self.token.get() {
+                    unsafe { self.inner.ni.encrypt_block(block) }
+                } else {
+                    unsafe { self.inner.soft.encrypt_block(block) }
+                }
+            }
+
+            #[inline]
+            fn encrypt_par_blocks(&self, blocks: &mut ParBlocks) {
+                if self.token.get() {
+                    unsafe { self.inner.ni.encrypt_par_blocks(blocks) }
+                } else {
+                    unsafe { self.inner.soft.encrypt_par_blocks(blocks) }
+                }
+            }
+        }
+
+        impl BlockDecrypt for $name {
+            #[inline]
+            fn decrypt_block(&self, block: &mut Block) {
+                if self.token.get() {
+                    unsafe { self.inner.ni.decrypt_block(block) }
+                } else {
+                    unsafe { self.inner.soft.decrypt_block(block) }
+                }
+            }
+
+            #[inline]
+            fn decrypt_par_blocks(&self, blocks: &mut ParBlocks) {
+                if self.token.get() {
+                    unsafe { self.inner.ni.decrypt_par_blocks(blocks) }
+                } else {
+                    unsafe { self.inner.soft.decrypt_par_blocks(blocks) }
+                }
+            }
+        }
+
+        opaque_debug::implement!($name);
+    }
+}
+
+define_aes_impl!(Aes128, aes128, U16, "AES-128 block cipher instance");
+define_aes_impl!(Aes192, aes192, U24, "AES-192 block cipher instance");
+define_aes_impl!(Aes256, aes256, U32, "AES-256 block cipher instance");
+
+#[cfg(feature = "ctr")]
+pub(crate) mod ctr {
+    use super::{Aes128, Aes192, Aes256};
+    use cipher::{
+        block::BlockCipher,
+        generic_array::GenericArray,
+        stream::{
+            FromBlockCipher, LoopError, OverflowError, SeekNum, SyncStreamCipher,
+            SyncStreamCipherSeek,
+        },
+    };
+
+    cpuid_bool::new!(aes_ssse3_cpuid, "aes", "ssse3");
+
+    macro_rules! define_aes_ctr_impl {
+        (
+            $name:tt,
+            $cipher:ident,
+            $module:tt,
+            $doc:expr
+        ) => {
+            #[doc=$doc]
+            #[cfg_attr(docsrs, doc(cfg(feature = "ctr")))]
+            pub struct $name {
+                inner: $module::Inner,
+            }
+
+            mod $module {
+                pub(super) enum Inner {
+                    Ni(crate::ni::$name),
+                    Soft(crate::soft::$name),
+                }
+            }
+
+            impl FromBlockCipher for $name {
+                type BlockCipher = $cipher;
+                type NonceSize = <$cipher as BlockCipher>::BlockSize;
+
+                fn from_block_cipher(
+                    cipher: $cipher,
+                    nonce: &GenericArray<u8, Self::NonceSize>,
+                ) -> Self {
+                    let (_, aesni_present) = aes_ssse3_cpuid::init_get();
```

Comment:
If you don't need the token, you can simply use `aes_ssse3_cpuid::get()`.

Reply:
Updated this to use `aes_cpuid::get()`, as per the Westmere-related comments AFAIK all AES-NI capable CPUs should have SSSE3.